File: backend/src/clients/deepgramClient.ts
import { createClient, DeepgramClient, LiveTranscriptionEvents } from '@deepgram/sdk';
import dotenv from 'dotenv';

dotenv.config();

const apiKey = process.env.DEEPGRAM_API_KEY;

let deepgramClient: DeepgramClient | null = null;

if (apiKey) {
  deepgramClient = createClient(apiKey);
  console.log('[Deepgram] Client initialized.');
} else {
  console.warn(
    'DEEPGRAM_API_KEY not found in environment variables. Deepgram functions will be disabled.'
  );
}

/**
 * Transcribes audio data using Deepgram.
 * @param audioBuffer - The audio data as a Buffer.
 * @returns A Promise resolving to the transcription text or null if failed.
 */
export const transcribeSpeech = async (
  audioBuffer: Buffer
): Promise<string | null> => {
  if (!deepgramClient) {
    console.warn('Deepgram client not initialized (API key missing?). Cannot transcribe.');
    return null;
  }

  console.log('[Deepgram] Received request to transcribe audio source.');

  try {
    const { result, error } = await deepgramClient.listen.prerecorded.transcribeFile(
      audioBuffer,
      {
        model: 'nova-2',
        smart_format: true,
        language: 'en-US',
      }
    );
    
    if (error) throw error;
    if (!result) throw new Error('No transcription result received.');
    
    const transcript = result.results.channels[0].alternatives[0].transcript;
    console.log(`[Deepgram] Transcription: ${transcript}`);
    return transcript;

  } catch (error) {
    console.error('[Deepgram] Error transcribing speech:', error);
    return null;
  }
};

/**
 * Transcribes audio from a URL using Deepgram.
 * @param audioUrl - The URL of the audio file to transcribe.
 * @returns A Promise resolving to the transcription text or null if failed.
 */
export const transcribeSpeechFromUrl = async (
  audioUrl: string
): Promise<string | null> => {
  if (!deepgramClient) {
    console.warn('Deepgram client not initialized (API key missing?). Cannot transcribe.');
    return null;
  }

  console.log(`[Deepgram] Received request to transcribe audio from URL: ${audioUrl}`);

  try {
    const { result, error } = await deepgramClient.listen.prerecorded.transcribeUrl(
      {
        url: audioUrl,
      },
      {
        model: 'nova-2',
        smart_format: true,
        language: 'en-US',
      }
    );
    
    if (error) throw error;
    if (!result) throw new Error('No transcription result received.');
    
    const transcript = result.results.channels[0].alternatives[0].transcript;
    console.log(`[Deepgram] Transcription from URL: ${transcript}`);
    return transcript;

  } catch (error) {
    console.error('[Deepgram] Error transcribing speech from URL:', error);
    return null;
  }
};

/**
 * Sets up a live transcription connection for real-time speech recognition.
 * @param options - Configuration options for the live transcription.
 * @param onTranscript - Callback function to handle transcript data.
 * @returns The live transcription connection object or null if initialization failed.
 */
export const setupLiveTranscription = (
  options: {
    language?: string;
    model?: string;
    encoding?: string;
    sampleRate?: number;
  } = {},
  onTranscript: (transcript: string) => void
) => {
  if (!deepgramClient) {
    console.warn('Deepgram client not initialized (API key missing?). Cannot setup live transcription.');
    return null;
  }

  const connection = deepgramClient.listen.live({
    model: options.model || 'nova-2',
    language: options.language || 'en-US',
    encoding: options.encoding || 'linear16',
    sampleRate: options.sampleRate || 16000,
    punctuate: true,
    smart_format: true,
  });

  connection.on(LiveTranscriptionEvents.Open, () => {
    console.log('[Deepgram Live] Connection opened.');
  });

  connection.on(LiveTranscriptionEvents.Transcript, (data) => {
    const transcript = data.channel.alternatives[0].transcript;
    if (transcript.trim()) {
      console.log('[Deepgram Live] Transcript received:', transcript);
      onTranscript(transcript);
    }
  });

  connection.on(LiveTranscriptionEvents.Close, () => {
    console.log('[Deepgram Live] Connection closed.');
  });

  connection.on(LiveTranscriptionEvents.Error, (error) => {
    console.error('[Deepgram Live] Error:', error);
  });

  return connection;
};       

File: backend/ambi/clients/deepgram_client.py
"""Deepgram API client for Ambi Python backend."""

import logging
from typing import Dict, Optional, Union

from deepgram import (
    DeepgramClient,
    DeepgramClientOptions,
    PrerecordedOptions,
)

from ambi.config.settings import DEEPGRAM_API_KEY

logger = logging.getLogger(__name__)


class DeepgramSpeechClient:
    """Client for interacting with Deepgram API for speech-to-text."""

    def __init__(self) -> None:
        """Initialize the Deepgram client."""
        if not DEEPGRAM_API_KEY:
            logger.error("DEEPGRAM_API_KEY is not defined in environment variables")
            self.client = None
        else:
            options = DeepgramClientOptions(api_key=DEEPGRAM_API_KEY)
            self.client = DeepgramClient(options)

    async def transcribe_audio(
        self,
        audio_data: Union[bytes, str],
        language: str = "en",
        model: str = "nova-2",
        smart_format: bool = True,
        diarize: bool = False,
    ) -> Optional[Dict]:
        """
        Transcribe audio data to text using Deepgram API.

        Args:
            audio_data: Audio data as bytes or file path as string
            language: Language code (default: 'en')
            model: Deepgram model to use (default: 'nova-2')
            smart_format: Whether to apply smart formatting (default: True)
            diarize: Whether to diarize the audio (default: False)

        Returns:
            Transcription result as dictionary or None if an error occurs
        """
        if not self.client:
            logger.error("Deepgram client is not initialized")
            return None

        try:
            options = PrerecordedOptions(
                model=model,
                language=language,
                smart_format=smart_format,
                diarize=diarize,
            )

            if isinstance(audio_data, bytes):
                response = await self.client.listen.prerecorded.transcribe_buffer(
                    buffer=audio_data,
                    options=options,
                )
            elif isinstance(audio_data, str):
                with open(audio_data, "rb") as audio_file:
                    response = await self.client.listen.prerecorded.transcribe_file(
                        file=audio_file,
                        options=options,
                    )
            else:
                logger.error("Invalid audio data type")
                return None

            return response.results

        except Exception as e:
            logger.error(f"Error transcribing audio with Deepgram: {str(e)}")
            return None

    def extract_transcript(self, result: Dict) -> str:
        """
        Extract transcript text from Deepgram response.

        Args:
            result: Deepgram transcription result

        Returns:
            Extracted transcript text
        """
        try:
            if not result or "channels" not in result:
                return ""

            channel = result["channels"][0]
            if "alternatives" not in channel or not channel["alternatives"]:
                return ""

            alternative = channel["alternatives"][0]
            if "transcript" not in alternative:
                return ""

            return alternative["transcript"]

        except Exception as e:
            logger.error(f"Error extracting transcript from Deepgram result: {str(e)}")
            return ""


deepgram_client = DeepgramSpeechClient()
